File: java/src/main/java/net/asam/openscenario/parser/ParserHelper.java
/*
 * Copyright 2020 RA Consulting
 *
 * RA Consulting GmbH licenses this file under the Apache License,
 * Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package net.asam.openscenario.parser;

import java.util.Date;

import javax.xml.bind.DatatypeConverter;

import net.asam.openscenario.common.ErrorLevel;
import net.asam.openscenario.common.FileContentMessage;
import net.asam.openscenario.common.IParserMessageLogger;
import net.asam.openscenario.common.Textmarker;

/**
 * Provides helper methods for the pasring process.
 *
 * @author Andreas Hege - RA Consulting
 */
public class ParserHelper {
  public static final int UNSIGNED_SHORT_MAX_VALUE = 2 * Short.MAX_VALUE + 1;

  public static final long UNSIGNED_INT_MAX_VALUE = 2L * (long) Integer.MAX_VALUE + 1L;

  /**
   * Parsing a string value into a string value.
   *
   * @param messageLogger to log errors and warnings
   * @param xmlValue the value from xml
   * @param textMarker text marker
   * @return the parsed value
   */
  public static String parseString(
      IParserMessageLogger messageLogger, String xmlValue, Textmarker textMarker) {
    return xmlValue;
  }

  /**
   * Parsing a string representation of an unsigned integer value into an unsigned integer value and
   * check the limits of unsigned integer (32 bit range). unsigned integer value is represented in a
   * Long data object.
   *
   * @param messageLogger to log errors and warnings
   * @param xmlValue the value from xml
   * @param textMarker text marker
   * @return the parsed value or null if value cannot be parsed
   */
  public static Long parseUnsignedInt(
      IParserMessageLogger messageLogger, String xmlValue, Textmarker textMarker) {

    try {
      Long result = DatatypeConverter.parseLong(xmlValue);
      if (result > UNSIGNED_INT_MAX_VALUE || result < 0) {
        messageLogger.logMessage(
            new FileContentMessage(
                "Cannot convert '"
                    + xmlValue
                    + "' to an unsignedInteger. Value must be in [0.."
                    + UNSIGNED_INT_MAX_VALUE
                    + "].",
                ErrorLevel.ERROR,
                textMarker));
      } else {
        return result;
      }

    } catch (NumberFormatException e) {
      messageLogger.logMessage(
          new FileContentMessage(
              "Cannot convert '" + xmlValue + "' to an unsignedInteger. Number format error.",
              ErrorLevel.ERROR,
              textMarker));
    }
    return null;
  }

  /**
   * Parsing a string representation of an integer value into an integer value. Integer value is
   * represented in a Integer data object.
   *
   * @param messageLogger to log errors and warnings
   * @param xmlValue the value from xml
   * @param textMarker text marker
   * @return the parsed value or null if value cannot be parsed
   */
  public static Integer parseInt(
      IParserMessageLogger messageLogger, String xmlValue, Textmarker textMarker) {

    try {
      Integer result = DatatypeConverter.parseInt(xmlValue);
      return result;

    } catch (NumberFormatException e) {
      messageLogger.logMessage(
          new FileContentMessage(
              "Cannot convert '" + xmlValue + "' to an int. Number format error.",
              ErrorLevel.ERROR,
              textMarker));
    }
    return null;
  }

  /**
   * Parsing a string representation of a double value into a double value. Integer value is
   * represented in a Double data object.
   *
   * @param messageLogger to log errors and warnings
   * @param xmlValue the value from xml
   * @param textMarker text marker
   * @return the parsed value or null if value cannot be parsed
   */
  public static Double parseDouble(
      IParserMessageLogger messageLogger, String xmlValue, Textmarker textMarker) {
    try {
      Double result = DatatypeConverter.parseDouble(xmlValue);
      return result;

    } catch (NumberFormatException e) {
      messageLogger.logMessage(
          new FileContentMessage(
              "Cannot convert '" + xmlValue + "' to a double. Number format error.",
              ErrorLevel.ERROR,
              textMarker));
    }
    return null;
  }

  /**
   * Parsing a string representation of an unsigned short value into an unsigned short value. and
   * check the limits of unsigned short (16 bit). unsigned short value is represented in a Integer
   * data object.
   *
   * @param messageLogger to log errors and warnings
   * @param xmlValue the value from xml
   * @param textMarker text marker
   * @return the parsed value or null if value cannot be parsed
   */
  public static Integer parseUnsignedShort(
      IParserMessageLogger messageLogger, String xmlValue, Textmarker textMarker) {

    try {
      Integer result = DatatypeConverter.parseInt(xmlValue);
      if (result > 2 * UNSIGNED_SHORT_MAX_VALUE || result < 0) {
        messageLogger.logMessage(
            new FileContentMessage(
                "Cannot convert '"
                    + xmlValue
                    + "' to an unsignedShort. Value must be in [0.."
                    + UNSIGNED_SHORT_MAX_VALUE
                    + "].",
                ErrorLevel.ERROR,
                textMarker));
      } else {
        return result;
      }

    } catch (NumberFormatException e) {
      messageLogger.logMessage(
          new FileContentMessage(
              "Cannot convert '" + xmlValue + "' to an unsignedShort. Number format error.",
              ErrorLevel.ERROR,
              textMarker));
    }
    return null;
  }

  /**
   * Parsing a string representation of a boolean into an boolean value. boolean value is
   * represented in a Boolean data object.
   *
   * @param messageLogger to log errors and warnings
   * @param xmlValue the value from xml
   * @param textMarker text marker
   * @return the parsed value or null if value cannot be parsed
   */
  public static Boolean parseBoolean(
      IParserMessageLogger messageLogger, String xmlValue, Textmarker textMarker) {

    if (xmlValue.equals("true") || xmlValue.equals("1")) {
      return true;
    } else if (xmlValue.equals("false") || xmlValue.equals("0")) {
      return false;
    }

    messageLogger.logMessage(
        new FileContentMessage(
            "Cannot convert '" + xmlValue + "' to a boolean. Illegal boolean value.",
            ErrorLevel.ERROR,
            textMarker));

    return null;
  }

  /**
   * Parsing a string representation of a date into a date value. Date value is represented in a
   * java.util.Date data object.
   *
   * @param messageLogger to log errors and warnings
   * @param xmlValue the value from xml
   * @param textMarker text marker
   * @return the parsed value or null if value cannot be parsed
   */
  public static Date parseDateTime(
      IParserMessageLogger messageLogger, String xmlValue, Textmarker textMarker) {

    try {
      Date result = DatatypeConverter.parseDateTime(xmlValue).getTime();
      return result;

    } catch (IllegalArgumentException e) {
      messageLogger.logMessage(
          new FileContentMessage(
              "Cannot convert '" + xmlValue + "' to a dateTime. Illegal dateTime value.",
              ErrorLevel.ERROR,
              textMarker));
    }
    return null;
  }

  /**
   * Parsing a string representation of an unsigned integer value into an unsigned integer value and
   * check the limits of unsigned integer (32 bit range). unsigned integer value is represented in a
   * Long data object.
   *
   * @param xmlValue the value
   * @throws Exception when validation fails
   */
  public static void validateUnsignedInt(String xmlValue) throws Exception {

    try {
      Long result = DatatypeConverter.parseLong(xmlValue);
      if (result > UNSIGNED_INT_MAX_VALUE || result < 0) {
        throw new Exception(
            "Cannot convert '"
                + xmlValue
                + "' to an unsignedInteger. Value must be in [0.."
                + UNSIGNED_INT_MAX_VALUE
                + "].");
      }

    } catch (NumberFormatException e) {
      throw new Exception(
          "Cannot convert '" + xmlValue + "' to an unsignedInteger. Number format error.");
    }
  }

  /**
   * Parsing a string representation of an integer value into an integer value. Integer value is
   * represented in a Integer data object.
   *
   * @param xmlValue the value
   * @throws Exception when validation fails
   */
  public static void validateInt(String xmlValue) throws Exception {

    try {
      DatatypeConverter.parseInt(xmlValue);

    } catch (NumberFormatException e) {
      throw new Exception("Cannot convert '" + xmlValue + "' to an int. Number format error.");
    }
  }

  /**
   * Parsing a string representation of a double value into a double value. Integer value is
   * represented in a Double data object.
   *
   * @param xmlValue the value
   * @throws Exception when validation fails
   */
  public static void validateDouble(String xmlValue) throws Exception {
    try {
      DatatypeConverter.parseDouble(xmlValue);

    } catch (NumberFormatException e) {
      throw new Exception("Cannot convert '" + xmlValue + "' to a double. Number format error.");
    }
  }

  /**
   * Parsing a string representation of an unsigned short value into an unsigned short value. and
   * check the limits of unsigned short (16 bit). unsigned short value is represented in a Integer
   * data object.
   *
   * @param xmlValue the value
   * @throws Exception when validation fails
   */
  public static void validateUnsignedShort(String xmlValue) throws Exception {

    try {
      Integer result = DatatypeConverter.parseInt(xmlValue);
      if (result > 2 * UNSIGNED_SHORT_MAX_VALUE || result < 0) {
        throw new Exception(
            "Cannot convert '"
                + xmlValue
                + "' to an unsignedShort. Value must be in [0.."
                + UNSIGNED_SHORT_MAX_VALUE
                + "].");
      }

    } catch (NumberFormatException e) {

      throw new Exception(
          "Cannot convert '" + xmlValue + "' to an unsignedShort. Number format error.");
    }
  }

  /**
   * Parsing a string representation of a boolean into an boolean value. boolean value is
   * represented in a Boolean data object.
   *
   * @param xmlValue the value
   * @throws Exception when validation fails
   */
  public static void validateBoolean(String xmlValue) throws Exception {

    if (!(xmlValue.equals("true")
        || xmlValue.equals("false")
        || xmlValue.equals("0")
        || xmlValue.equals("1")))
      throw new Exception("Cannot convert '" + xmlValue + "' to a boolean. Illegal boolean value.");
  }

  /**
   * Parsing a string representation of a date into a date value. Date value is represented in a
   * java.util.Date data object.
   *
   * @param xmlValue the value
   * @throws Exception when validation fails
   */
  public static void validateDateTime(String xmlValue) throws Exception {

    try {
      Date result = DatatypeConverter.parseDateTime(xmlValue).getTime();

    } catch (IllegalArgumentException e) {
      throw new Exception(
          "Cannot convert '" + xmlValue + "' to a dateTime. Illegal dateTime value.");
    }
  }
}

File: cpp/openScenarioLib/src/parser/ParserHelper.cpp
/*
 * Copyright 2020 RA Consulting
 *
 * RA Consulting GmbH licenses this file under the Apache License,
 * Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#pragma once

#include  "ParserHelper.h"


/**
 * Container for parameter values (name, type, value)
 * @author Andreas Hege - RA Consulting
 *
 */
namespace NET_ASAM_OPENSCENARIO
{
    std::string ParserHelper::ParseString(IParserMessageLogger& messageLogger, std::string& xmlValue, Textmarker& textMarker)
    {
        return xmlValue;
    }

   uint32_t ParserHelper::ParseUnsignedInt(IParserMessageLogger& messageLogger, std::string& xmlValue, Textmarker& textMarker)
    {
        try 
        {
            const auto kResult = std::stoll(xmlValue, nullptr, 0);
            if (kResult > UNSIGNED_INT_MAX_VALUE || kResult < 0)
            {
                auto msg = FileContentMessage("Cannot convert '" + xmlValue + "' to an unsignedInteger. Value must be in [0.." + 
                                              std::to_string(UNSIGNED_INT_MAX_VALUE) + "].", ERROR, textMarker);
                messageLogger.LogMessage(msg);
            }
            else 
                return kResult & 0xffffffff;

        }
        catch ( std::invalid_argument& e ) 
        {
            (void)e;
            auto msg = FileContentMessage("Cannot convert '" + xmlValue + "' to an unsignedInteger. Number format error.", ERROR, textMarker);
            messageLogger.LogMessage(msg);
        }

        return 0;
    }

    int ParserHelper::ParseInt(IParserMessageLogger& messageLogger, std::string& xmlValue, Textmarker& textMarker)
    {
        try 
        {
            const auto kResult = std::stoi(xmlValue, nullptr, 0);
            return kResult;
        }
        catch (std::invalid_argument& e)
        {
            (void)e;
            auto msg = FileContentMessage("Cannot convert '" + xmlValue + "' to an int. Number format error.", ERROR, textMarker);
            messageLogger.LogMessage(msg);
        }
        return 0;
    }

    double ParserHelper::ParseDouble(IParserMessageLogger& messageLogger, std::string& xmlValue, Textmarker& textMarker)
    {
        try 
        {
            const auto kResult = std::stod(xmlValue);
            return kResult;

        }
        catch (std::invalid_argument& e) 
        {
            (void)e;
            auto msg = FileContentMessage("Cannot convert '" + xmlValue + "' to a double. Number format error.", ERROR, textMarker);
            messageLogger.LogMessage(msg);
        }
        return 0;
    }

   uint16_t ParserHelper::ParseUnsignedShort(IParserMessageLogger& messageLogger, std::string& xmlValue, Textmarker& textMarker)
    {
        try 
        {
            const auto kResult = std::stol(xmlValue, nullptr, 0);
            if (kResult > 2 * UNSIGNED_SHORT_MAX_VALUE || kResult < 0)
            {
                auto msg = FileContentMessage("Cannot convert '" + xmlValue + "' to an unsignedShort. Value must be in [0.." +
                                               std::to_string(UNSIGNED_SHORT_MAX_VALUE) + "].", ERROR, textMarker);
                messageLogger.LogMessage(msg);
            }
            else 
                return kResult & 0xffff;

        }
        catch (std::invalid_argument& e) 
        {
            (void)e;
            auto msg = FileContentMessage("Cannot convert '" + xmlValue + "' to an unsignedShort. Number format error.", ERROR, textMarker);
            messageLogger.LogMessage(msg);
        }
        return 0;
    }

    bool ParserHelper::ParseBoolean(IParserMessageLogger& messageLogger, std::string& xmlValue, Textmarker& textMarker)
    {
        bool result;

//            std::transform(xmlValue.begin(), xmlValue.end(), xmlValue.begin(), [](unsigned char c) { return std::tolower(c, std::locale()); });
        for ( std::string::iterator it=xmlValue.begin(); it!=xmlValue.end(); ++it)
           *it = std::tolower(*it, std::locale());

        if (xmlValue == "1" || xmlValue == "true")
            result = true;
        else if (xmlValue == "0" || xmlValue == "false")
            result = false;
        else
        {
            auto msg = FileContentMessage("Cannot convert '" + xmlValue + "' to a boolean. Number format error.", ERROR, textMarker);
            messageLogger.LogMessage(msg);
            return false;
        }

        return result;
    }

   DateTime ParserHelper::ParseDateTime(IParserMessageLogger& messageLogger, std::string& xmlValue, Textmarker& textMarker)
    {
        DateTime result{};

        try
        {
            if(!DateTimeParser::ToDateTime(xmlValue, result))
            {
                auto msg = FileContentMessage("Cannot convert '" + xmlValue + "' to a dateTime. Number format error.", ERROR, textMarker);
                messageLogger.LogMessage(msg);
                return {};
            }
        }
        catch (...)
        {
            auto msg = FileContentMessage("Cannot convert '" + xmlValue + "' to a dateTime. Number format error.", ERROR, textMarker);
            messageLogger.LogMessage(msg);
            return {};
        }
        return result;
    }

    void ParserHelper::ValidateUnsignedInt(std::string& xmlValue)
    {
        try 
        {
            const auto kResult = std::stoll(xmlValue, nullptr, 0);

            if (kResult > UNSIGNED_INT_MAX_VALUE || kResult < 0)
            {
                const std::string kErrorMessage = "Cannot convert '" + xmlValue + "' to an unsignedInteger. Value must be in [0.." + std::to_string(UNSIGNED_INT_MAX_VALUE)+ "].";
                throw std::range_error(kErrorMessage);
            }

        }
        catch (...) 
        {
            throw std::runtime_error("Cannot convert '" + xmlValue + "' to an unsignedInteger. Number format error.");
        }
    }

    void ParserHelper::ValidateInt(std::string& xmlValue)
    {
        try 
        {
            (void) std::stoi(xmlValue, nullptr, 0);
        }
        catch (...) 
        {
            throw std::runtime_error("Cannot convert '" + xmlValue + "' to an int. Number format error.");
        }
    }

    void ParserHelper::ValidateDouble(std::string& xmlValue)
    {
        try 
        {
            (void) std::stod(xmlValue);
        }
        catch (...)
        {
            throw std::runtime_error("Cannot convert '" + xmlValue + "' to a double. Number format error.");
        }
    }

    void ParserHelper::ValidateUnsignedShort(std::string& xmlValue)
    {
        try 
        {
            const auto kResult = std::stol(xmlValue, nullptr, 0);
            if (kResult > 2 * UNSIGNED_SHORT_MAX_VALUE || kResult < 0)
            {
                const std::string kErrorMessage = "Cannot convert '" + xmlValue + "' to an unsignedShort. Value must be in [0.." + std::to_string(UNSIGNED_SHORT_MAX_VALUE) + "].";
                throw std::range_error(kErrorMessage);
            }
        }
        catch (...)
        {
            throw std::runtime_error("Cannot convert '" + xmlValue + "' to an unsignedShort. Number format error.");
        }
    }

    void ParserHelper::ValidateBoolean(std::string& xmlValue)
    {
        auto xmlValueCpy = xmlValue;
        for (std::string::iterator it = xmlValueCpy.begin(); it != xmlValueCpy.end(); ++it)
            *it = std::tolower(*it, std::locale());

        if (xmlValueCpy !="true" && xmlValueCpy != "false" && xmlValueCpy != "0" && xmlValueCpy != "1")
            throw std::runtime_error("Cannot convert '" + xmlValue + "' to a boolean. Illegal boolean value.");
    }

    void ParserHelper::ValidateDateTime(std::string& xmlValue)
    {
        try 
        {
            DateTime result{};
            if (!DateTimeParser::ToDateTime(xmlValue, result))
            {
                if (xmlValue != "true" && xmlValue != "false" && xmlValue != "0" && xmlValue != "1")
                    throw std::runtime_error("Cannot convert '" + xmlValue + "' to a dateTime. Illegal dateTime value.");
            }
        }
        catch (...)
        {
            throw std::runtime_error("Cannot convert '" + xmlValue + "' to a dateTime. Illegal dateTime value.");
        }
    }

}
